### PR TITLE
Migrate sharded ops from TTL to TTNN

### PR DESCRIPTION
**Pull request by ntarafdar**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/10131)

### Problem description
Migrating sharded ops (interleaved_to_sharded, sharded_to_interleaved, and reshard)

### What's changed
Migrated and replaced uses

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by djordje-tt** [comment]
Please, make sure next T3K pipelines are passing: unit tests, frequent tests, demo, model perf pipelines.

**Review by eyonland — APPROVED** [review]

**Review by cglagovichTT — APPROVED** [review]
Llama ✅

**Review by esmalTT — APPROVED** [review]
SD+Mamba LGTM 👍

**Review by tt-aho — APPROVED** [review]
Bert + Resnet look okay.

Please review the license header on the new files that were added. They should all use the current year but some of them still use 2023.

**Review by tt-rkim — APPROVED** [review]

File: models/demos/falcon7b_common/tt/falcon_attention.py [hunk 1]
```diff
@@ -716,7 +716,7 @@ def forward(
 
         if self.model_config["l1_sharded"]:
             for i in range(self.num_devices):
-                key_layer[i] = ttnn.experimental.tensor.interleaved_to_sharded(
+                key_layer[i] = ttnn.interleaved_to_sharded(
                     key_layer[i],
                     sharded_mem_config=self.model_config["ATTN_BATCH_SHARDED_MEMCFG"](
```
Comment (skhorasganiTT) [inline models/demos/falcon7b_common/tt/falcon_attention.py:721]: sharded_memory_config?

File: models/demos/falcon7b_common/tt/falcon_attention.py [hunk 2]
```diff
@@ -749,7 +749,7 @@ def forward(
                 )
 
             for i in range(self.num_devices):
-                query_layer[i] = ttnn.experimental.tensor.interleaved_to_sharded(
+                query_layer[i] = ttnn.interleaved_to_sharded(
                     query_layer[i],
                     sharded_mem_config=self.model_config["ATTN_BATCH_SHARDED_MEMCFG"](
```
Comment (skhorasganiTT) [inline models/demos/falcon7b_common/tt/falcon_attention.py:754]: sharded_memory_config?

File: models/demos/falcon7b_common/tt/falcon_attention.py [hunk 3]
```diff
@@ -879,7 +879,7 @@ def forward(
             )
         if self.model_config["l1_sharded"]:
             for i in range(self.num_devices):
-                value_layer[i] = ttnn.experimental.tensor.interleaved_to_sharded(
+                value_layer[i] = ttnn.interleaved_to_sharded(
                     value_layer[i],
                     sharded_mem_config=self.model_config["ATTN_BATCH_SHARDED_MEMCFG"](
```
Comment (skhorasganiTT) [inline models/demos/falcon7b_common/tt/falcon_attention.py:884]: sharded_memory_config?

File: models/demos/falcon7b_common/tt/falcon_attention.py [hunk 4]
```diff
@@ -915,7 +915,7 @@ def forward(
                 value_layer[i].deallocate()
 
             for i in range(self.num_devices):
-                attn_output[i] = ttnn.experimental.tensor.sharded_to_interleaved(
+                attn_output[i] = ttnn.sharded_to_interleaved(
                     attn_output[i], output_mem_config=self.model_config["POST_SOFTMAX_MM_OUTPUT_MEMCFG"]
```
Comment (skhorasganiTT) [inline models/demos/falcon7b_common/tt/falcon_attention.py:919]: is output_mem_config still valid?

File: models/demos/falcon7b_common/tt/model_utils.py [hunk 5]
```diff
@@ -96,7 +96,7 @@ def layernorm(ln_input, ln_eps, ln_gamma, ln_betta, num_devices, model_config):
     if h_dim in [32, 128, 256, 1024, 2048]:
         for i in range(num_devices):
             ln_output.append(
-                ttnn.experimental.tensor.interleaved_to_sharded(
+                ttnn.interleaved_to_sharded(
                     ln_input[i], sharded_mem_config=model_config["LAYERNORM_BLOCK_SHARDED_MEM_CFG"][h_dim]
```
Comment (skhorasganiTT) [inline models/demos/falcon7b_common/tt/model_utils.py:100]: sharded_memory_config?

**Comment by tarafdarTT** [comment]
thanks @skhorasganiTT ya missed all the mem config issues, I updated that.

**Review by skhorasganiTT — APPROVED** [review]

**Review by sjameelTT — APPROVED** [review]

**Review by johanna-rock-tt — APPROVED** [review]
Falcon40b changes look good to me. Please make sure to run all t3k pipelines.

**Review by yieldthought — APPROVED** [review]

**Review by mywoodstock — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/core/to_memory_config/to_memory_config_op.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/core/to_memory_config/to_memory_config_op.hpp:1]: These core ops should also go to data movement I think

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/core/to_memory_config/to_memory_config_op.hpp:1]: not in this PR, but in some next :D

File: ttnn/cpp/ttnn/operations/data_movement/sharded/common.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/sharded/common.hpp:1]: Please rename to `sharded_common.hpp` or `shareded_utils.hpp`. Avoid generic names like "common.hpp" or "types.hpp"

File: ttnn/cpp/ttnn/operations/data_movement/sharded/interleaved_to_sharded/device/interleaved_to_sharded_op.hpp [hunk 6]
```diff
@@ -0,0 +1,32 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "ttnn/tensor/tensor.hpp"
+#include "ttnn/run_operation.hpp"
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/sharded/interleaved_to_sharded/device/interleaved_to_sharded_op.hpp:8]: can remove this include I think

File: ttnn/cpp/ttnn/operations/data_movement/sharded/reshard/device/reshard_op.hpp [hunk 7]
```diff
@@ -0,0 +1,30 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <optional>
+
+#include "ttnn/tensor/tensor.hpp"
+#include "ttnn/run_operation.hpp"
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/sharded/reshard/device/reshard_op.hpp:10]: please remove, heavy include

File: ttnn/cpp/ttnn/operations/data_movement/sharded/sharded_to_interleaved/device/sharded_to_interleaved_op.hpp [hunk 8]
```diff
@@ -0,0 +1,32 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "ttnn/tensor/tensor.hpp"
+#include "ttnn/run_operation.hpp"
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/sharded/sharded_to_interleaved/device/sharded_to_interleaved_op.hpp:8]: should be able to remove

**Review by ayerofieiev-tt — APPROVED** [review]
Great job! Thank you!

File: ttnn/cpp/ttnn/operations/core/to_memory_config/to_memory_config_op.hpp
Comment (tarafdarTT) [reply ttnn/cpp/ttnn/operations/core/to_memory_config/to_memory_config_op.hpp:1]: ah yes that's a good idea!

**Pull request merged**